lib/tax_tribunal/login.rb

```diff
@@ -9,24 +9,30 @@ class Login < Downloader
 
     get '/login' do
       if session[:email].nil?
+        logger.info({ action: 'login', state: 'new', message: 'new login', requested: session[:return_to] })
         redirect oauth_client.auth_code.authorize_url(redirect_uri: CALLBACK_URI)
       else
+        logger.info({ action: 'login', state: 'existing', message: session[:email] })
         session[:already_logged_in] = "You are already logged in as #{session[:email]}."
         erb :root
       end
     end
 
     get '/logout' do
+      logger.info({ action: 'logout', message: session[:email] })
       session.destroy
       erb :root
     end
 
     get '/oauth/callback' do
       resp = authorise!(params[:code])
-      if resp
+      links = resp.fetch(:links, {})
+      if resp.fetch(:email, nil) && links.fetch(:logout, nil) && links.fetch(:profile, nil)
```

Comment:
I'd be inclined to default to "false" instead of "nil" here, to make the intention clear